src/libstd/sync/mutex.rs

```diff
@@ -358,6 +358,37 @@ impl<T: ?Sized> Mutex<T> {
         let data = unsafe { &mut *self.data.get() };
         poison::map_result(self.poison.borrow(), |_| data )
     }
+
+    /// Invoke a function under a mutex.
+    ///
+    /// This takes the mutex and once held passes a mutable reference to the
+    /// contained value to a `FnOnce` closure. Once the closure returns, the
+    /// mutex is released.
+    ///
+    /// # Errors
+    ///
+    /// If another user of this mutex panicked while holding the mutex, then
+    /// this call will return an error instead.
+    ///
+    /// # Example
+    /// ```
+    /// use std::sync::Mutex;
+    ///
+    /// let mutex = Mutex::new(0);
+    ///
+    /// // Atomically fetch the old value and increment it.
+    /// let old = mutex.with(|v| { let old = v; *v += 1; old }).unwrap();
+    ///
+    /// assert_eq!(old, 0);
+    /// assert_eq!(*mutex.lock().unwrap(), 1);
+    /// ```
+    #[unstable(feature = "mutex_with", issue = "61974")]
+    pub fn with<U>(
+        &self,
+        func: impl FnOnce(&mut T) -> U,
+    ) -> Result<U, PoisonError<MutexGuard<'_, T>>> {
```

Comment:
Why would the PoisonError contain the guard in this case?

Reply:
I was making it as similar to `LockResult` as possible, with the same error type. I guess the rationale is that if you have a poisoned lock, and you can actually recover from that, then having the guard would allow you to do so. On the other hand, that's at odds with point of this API, which is to be simple and low fuss.

I think the alternatives are:
1. Use `PoisonError<()>`
2. Don't return `Result` at all, and just panic on a poisoned lock

Since I think poisoned locks are unrecoverable in practice, 2. is possibly justifiable, but it seems a bit rude for a libstd function to do that. And as far as 1. goes, if we're going to return an error anyway, it may as well have the guard in it, even if its never used in practice.

Reply:
We could add a `try_with` variant for the non-panicing version.

Reply:
So `with` panics, `try_with` returns a `Result`?

Reply:
Yeah that's the idea.

Reply:
I updated to include `try_with`, but TBH I think it is overkill. I don't think this
```
match mx.with(|v| *v += 1234) {
Ok(()) => (),
Err(err) => ... handle err ...,
```
is any improvement over this:
```
match mx.lock() {
Ok(lk) => *lk += 1234,
Err(err) => ... handle err ...,
}
```